addons/marketing_card/models/card_campaign.py

```diff
@@ -195,62 +287,58 @@ def _get_image_b64(self, record):
-                    'header': _('Title'),
-                    'subheader': _('Subtitle'),
-                }
-            },
-            raise_on_code=False,
-        )
-        image_bytes = self.env['ir.actions.report']._run_wkhtmltoimage(
-            [rendered_body],
-            *TEMPLATE_DIMENSIONS
-        )[0]
-        return image_bytes and base64.b64encode(image_bytes)
-
     # ==========================================================================
     # Card creation
     # ==========================================================================
 
-    def _generate_card_hash_token(self, record_id):
-        """Generate a token for a specific recipient of this campaign."""
-        self.ensure_one()
-        return hmac(self.env(su=True), ('marketing_card', self._name, self._origin.id, self.name), record_id)
-
-    def _get_or_create_cards_from_res_ids(self, res_ids):
-        """Create missing cards for the given ids."""
+    def _update_cards(self, domain, auto_commit=False):
+        """Create missing cards and update cards if necessary based for the domain."""
         self.ensure_one()
-        cards = self.env['card.card'].search_fetch([('campaign_id', '=', self.id), ('res_id', 'in', res_ids)], ['res_id'])
-        missing_ids = set(res_ids) - set(cards.mapped('res_id'))
-        cards += self.env['card.card'].create([{'campaign_id': self.id, 'res_id': missing_id} for missing_id in missing_ids])
-
-        # order based on input
-        res_order = dict(zip(res_ids, count()))
-        return cards.sorted(key=lambda card: res_order[card.res_id])
-
-    def _get_preview_url_from_res_id(self, res_id):
-        return self._get_card_path(res_id, 'preview')
+        TargetModel = self.env[self.res_model]
+        res_ids = TargetModel.search(domain).ids
+        cards = self.env['card.card'].search_fetch([
+            ('active', 'in', [False, True]),
+            ('campaign_id', '=', self.id),
+            ('res_id', 'in', res_ids),
+        ], ['res_id', 'requires_sync'])
+        # update active and res_model for preview cards
+        cards.active = True
+        self.env['card.card'].create([
+            {'campaign_id': self.id, 'res_id': res_id}
+            for res_id in set(res_ids) - set(cards.mapped('res_id'))
+        ])
 
-    def _get_card_path(self, res_id, suffix):
-        self.ensure_one()
-        return f'{self.get_base_url()}/cards/{self._origin.id}/{res_id}/{self._generate_card_hash_token(res_id)}/{suffix}'
+        # render by batch of 100 to avoid losing progress in case of time out
+        offset = 0
+        updated_cards = self.env['card.card']
+        while cards := self.env['card.card'].search_fetch([
+            ('requires_sync', '=', True),
+            ('campaign_id', '=', self.id),
+            ('res_id', 'in', res_ids),
+        ], ['res_id'], limit=100, offset=offset):
+            # no need to autocommit if it can be done in one batch
+            if auto_commit and offset > 0:
```

Comment:
Maybe we could invalidate_cache at the end of the loop in order to free memory of the images